### PR TITLE
[Fizz] Implement New Context

### PR DESCRIPTION
**Pull request by sebmarkbage**
Fizz is a bit different than either Fiber or the partial renderer in terms of scenarios and tradeoffs. Unlike the partial renderer, we don't just always resume where we previously left off. We can suspend in different sibling trees. Unlike Fiber, we never have to restart with new props and we don't otherwise need the tree we've already past.

We could use a model similar to partial renderer's use of threads, where each context has N number of concurrent threads and a thread is allocated for each suspended task. But that would require a large and growing array for each Context which probably doesn't end up paying for itself.

We could use a Map of contexts associated with each provider. This requires a lot of cloning at the point of each provider. Effectively what legacy context does.

We could also store a loop up list of all parent providers that we backtrack any time we read a Context. That comes at the cost of reading being slower.

The general tradeoffs that we want to favor is making reading cheap so that mostly global dependency injection things are very cheap. We also generally favor rendering at full perf once something is unsuspended. Even though there are many possible suspended points, if you're preload a lot of data then you hopefully can just render straight through large sections without suspending at all.

The approach that I actually went with is similar to the Fiber approach in that we update the `_currentValue` field on each context and then reading is as cheap as just reading that value.

I also store a linked list of the stack of context providers that we're currently in. This gets heap allocated each time we visit a provider. If we need to spawn a new task, this node is what gets stored on the Task.

When we resume work on a task, we switch context by popping and pushing the delta between the most recently active context and the task's context. I.e. we restore where we were.

The cost of this happens at the point when we switch between tasks. It's cheap (or free) when we're resuming a single task over and over. Such as when the first few root components suspend or when you've reach the last suspended subtree which is the final one blocking completion. It's slower when you're switching back and forth between two deep and far removed siblings but at that point you're not close to reaching completion anyway.

_(This could use some more tests of various suspending and resuming scenarios that are specific to Fizz and isn't covered by existing tests. The coverage isn't great yet. But didn't want to spend too much time on it before I unblock other work on top.)_

### CODE REVIEW

File: packages/react-server/src/ReactFizzNewContext.js [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+// At the cost of updating contexts even if they're never read by this subtree.
+export function switchContext(newSnapshot: ContextSnapshot): void {
+  // The basic algorithm we need to do is to pop back any contexts that are no longer on the stack.
+  // We also need to update any new contexts that are now on the stack with the deepest value.
+  // The easiest way to update new contexts is to just reapply them in reverse order from the
+  // perspective of the backpointers. To avoid allocating a lot when switching, we use the stack
+  // for that. Therefore this algorithm is recursive.
+  // 1) First we pop which ever snapshot tree was deepest. Popping old contexts as we go.
+  // 2) Then we find the nearest common ancestor from there. Popping old contexts as we go.
+  // 3) Then we reapply new contexts on the way back up the stack.
```
Comment (sebmarkbage) [inline packages/react-server/src/ReactFizzNewContext.js:152]: @sophiebits You might enjoy this algorithm.

I also considered only storing the "value" and not depth/previousValue. That can be used to create a mode that always pushes all contexts - unless the new context is a strict superset. That could make the whole thing a bit lighter and still be fast when it only unsuspends in strict parent->child order.

But I think the pure common ancestor approach is probably worth it.

File: packages/react-server/src/ReactFizzServer.js [hunk 2]
```diff
@@ -1148,6 +1319,16 @@ function performWork(request: Request): void {
     fatalError(request, error);
   } finally {
     ReactCurrentDispatcher.current = prevDispatcher;
+    if (prevDispatcher === Dispatcher) {
```
Comment (sebmarkbage) [inline packages/react-server/src/ReactFizzServer.js:1322]: This case is meant to handle something like:

```
function Foo() {
  renderToString(<Bar />);
}
renderToString(<Foo />);
```

In a follow up.

File: packages/react-dom/src/server/ReactDOMServerFormatConfig.js [hunk 3]
```diff
@@ -48,6 +48,10 @@ import hasOwnProperty from 'shared/hasOwnProperty';
 import sanitizeURL from '../shared/sanitizeURL';
 import isArray from 'shared/isArray';
 
+// Used to distinguish these contexts from ones used in other renderers.
+// E.g. this can be used to distinguish legacy renderers from this modern one.
+export const isPrimaryRenderer = true;
```
Comment (sebmarkbage) [inline packages/react-dom/src/server/ReactDOMServerFormatConfig.js:53]: I plan on building renderToString and Fizz in separate renderers so this will be used to deal with this case:

```js
function Foo() {
  renderToString(<Bar />);
}
renderToWritable(<Foo />);
```

File: packages/react-server/src/ReactFizzNewContext.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    parentValue: prevValue,
+    value: nextValue,
+  };
+  currentActiveSnapshot = newNode;
+  return newNode;
+}
+
+export function popProvider<T>(context: ReactContext<T>): void {
+  const prevSnapshot = currentActiveSnapshot;
+  invariant(
```
Comment (gaearon) [inline packages/react-server/src/ReactFizzNewContext.js:224]: Why check this in prod?

File: packages/react-server/src/ReactFizzNewContext.js [hunk 5]
```diff
@@ -41,9 +41,133 @@ export const rootContextSnapshot: ContextSnapshot = null;
...
+  const prev = currentActiveSnapshot;
+  const next = newSnapshot;
+  if (prev !== next) {
+    if (prev === null) {
+      // $FlowFixMe: This has to be non-null since it's not equal to prev.
+      pushAllNext(next);
+    } else if (next === null) {
+      popAllPrevious(prev);
+    } else if (prev.depth === next.depth) {
+      popToNearestCommonAncestor(prev, next);
```
Comment (gaearon) [inline packages/react-server/src/ReactFizzNewContext.js:162]: The naming of this and the next one confused me because they don't only pop, but also push the missing ones.

File: packages/react-server/src/ReactFizzNewContext.js [hunk 6]
```diff
@@ -41,9 +41,133 @@ export const rootContextSnapshot: ContextSnapshot = null;
...
+    'The depth must equal at least at zero before reaching the root. This is a bug in React.',
+  );
+  if (prev.depth === parentNext.depth) {
+    // We found the same level. Now we just need to find a shared ancestor.
+    popToNearestCommonAncestor(prev, parentNext);
+  } else {
+    // We must still be deeper.
+    popNextToCommonLevel(prev, parentNext);
+  }
+  pushNode(next);
```
Comment (gaearon) [inline packages/react-server/src/ReactFizzNewContext.js:137]: Why is this needed? Same question re: initial pop in the previous method.

**Review by gaearon — APPROVED** [review]

File: packages/react-server/src/ReactFizzNewContext.js (hunk 4 above)
Comment (sebmarkbage) [reply packages/react-server/src/ReactFizzNewContext.js:224]: It follows the principle that if it's getting checked by runtime type checks here anyway, it likely isn't a perf cost (other than the extra code). The `prevSnapshot.context` access after does that anyway.

File: packages/react-server/src/ReactFizzNewContext.js (hunk 5 above)
Comment (sebmarkbage) [reply packages/react-server/src/ReactFizzNewContext.js:162]: Yea I thought that when I was naming them. Not sure what to call it. Maybe `...AndThenPush`.

File: packages/react-server/src/ReactFizzNewContext.js (hunk 6 above)
Comment (sebmarkbage) [reply packages/react-server/src/ReactFizzNewContext.js:137]: In the case one is deeper than the other there won't be the same amount of pop/push calls. If the previous is deeper, we need to pop more times than we push. If the next is deeper, we need to push more than we pop.
